toolchain/sem_ir/decl.h

```diff
@@ -0,0 +1,23 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_SEM_IR_DECL_H_
+#define CARBON_TOOLCHAIN_SEM_IR_DECL_H_
+
+#include "toolchain/sem_ir/ids.h"
+
+namespace Carbon::SemIR {
+
+// Information about a declaration.
+struct DeclInfo {
+  // The pattern block, containing the pattern insts for the parameter pattern.
+  InstBlockId pattern_block_id;
+  // The declaration block, containing the declaration's parameters and their
+  // types.
+  InstBlockId decl_block_id;
+};
```

Comment:
> note that `Impl` doesn't derive from `EntityWithParams` (yet?),

Note, https://github.com/carbon-language/carbon-lang/pull/4336 is changing this.

Reply:
> > Side note: this changes the representation of chains of function-style qualifiers like `Class(T:! type).F(n: T)` (see `toolchain/check/testdata/class/generic/member_out_of_line.carbon` for an example), but I think both forms are probably wrong, and I think the pre-existing representation of the decl is probably wrong too, so I'd rather punt on that whole issue for now.
> 
> I'm not seeing the issue, can you elaborate?

See `%F.decl` in that file: after my revisions last week (specifically [this one](https://github.com/carbon-language/carbon-lang/commit/ed374dcd7f3aad4d6c2b7acf092ae931d9c5742a#diff-5d84bb40fe60cd6c7ee3ddd37fcb93c1d9027118b6fa20f1896f539301b1404bR149)), there's no longer a `symbolic_binding_pattern T` in the pattern block. By contrast, there _is_ a `bind_symbolic_name T` in the decl block.

I'm going to take your other questions in reverse order:

> 2. Considering programmatic access for tools, how would a tool disambiguate between explicit and implicit parameter patterns?

My plan is to extend `NameComponent` and `EntityWithParams` with two new `InstBlockId` members that relate to the pattern block in much the same way that `implicit_params_id` and `params_id` currently relate to the decl block:
- They don't "own" the insts they contain, but refer to insts in the pattern block.
- They have one inst per parameter.
- Each inst is the "root" inst for the corresponding parameter: all the pattern insts relating to that parameter are reachable from it.
This isn't just for future SemIR consumers; `check` itself is going to need it once we shift to generating the pattern-match SemIR from the pattern SemIR.

I already have that implemented, but it seems out of scope for this PR.

> 1. Why do declarations always have a pattern block, even declarations that never have a pattern (such as alias and namespace)?

The basic problem is that the AST doesn't give us a reliable context-free signal that a full-pattern is starting (or ending), so we have to push a pattern block whenever one _might_ be about to start, and pop once the full-pattern (if any) has definitely ended. That applies even to `alias` and `namespace` because the AST allows those to be parameterized.

You're right that we could avoid that problem if we put the implicit and explicit parameter lists in separate pattern blocks, and used separate AST node kinds for parameter lists and other tuple patterns. However, the consensus [on Discord](https://discord.com/channels/655572317891461132/655578254970716160/1280995814805737513) seemed to be to have one pattern block for each full-pattern, and to me that seems like the more semantically faithful representation.

Furthermore, I don't know if that change would really gain us much. My strong suspicion is that if a SemIR consumer cares about the distinction between implicit and explicit parameters, they probably care about which parameter a given inst is associated with. That means they aren't going to want to work with the pattern block directly, because it's too unstructured -- they're going to want to work with the one-inst-per-parameter blocks that I mentioned earlier, and traverse from there. That's certainly been my experience so far with consuming the pattern IR in toolchain code.